pkg/controller/provider/web/openstack/base.go

```diff
@@ -79,3 +83,23 @@ func (r *PathBuilder) Path(m model.Model) (path string) {
 
 	return
 }
+
+// Path based on Project.
+func (r *PathBuilder) forProject(id, leaf string) (path string, err error) {
```

Comment:
can it be that id is different than tenantId or leaf is different than a vm name? it's written in a generic way

Reply:
Could be, for example, in the future we might want to add the project networks to the tree:
```go
case *model.Network:
		path = m.(*model.Network).Name
		object := m.(*model.Network)
		path, err = r.forProject(object.ProjectID, object.Name)
```
In this case `id` is `ProjectID` and `leaf` is `Name` but it could be any other object field like object.ID.
Not sure if I am answering your question.

Reply:
sort of :)
so if the 'id' is that of the project then let's call it projectId and would subresource be a better fit for the second argument?

Reply:
What about `childName` instead of `subresource`?

Reply:
sounds good

Reply:
Hold on, After reviewing @yaacov's PR I realized the projects are organized in a tree hierarchy.